### PR TITLE
Bounce-acccount custom fetchmail configuration and encryption

### PR DESCRIPTION
**Pull request by marvin-martian**
Hi,

I have modified the bounce-accounts to allow for a custom fetchmail-configuration field and a port field. I have also added encryption of both bounce-account passwords and the fetchmail-configuration in db. I encrypted the fetchmail config because it would also be possible to also add passwords in it.

This pull request solves issues #68, #11 

The fetchmail-configuration supports markers:
###SERVER###
###PROTOCOL###
###PORT###
###USERNAME###
###PASSWORD###

Old configurations are updated on install so they will not break.
Maybe I should have split this as 2 issues, but I was on a roll, sorry for that.

Have a nice day.

### CODE REVIEW

**Comment by PowerKiKi** [comment]
This sounds great ! But I won't have time to check the details before next week. Meanwhile could you re-format the PR ? this project does not use TYPO3 coding guideline, but instead PSR. The easiest way to do that is to build the project:
```
npm install && npm install -g gulp-cli
gulp
```
The build will fail, but you'll have php-cs-fixer installed. To actually fix it, run:
```
./vendor/bin/php-cs-fixer fix .
```

**Comment by marvin-martian** [comment]
OK you have your PSR corrections.. 

..just had a thought, this patch would create a dependency that the mcrypt php module is installed on the server before updating the extension.  I have made no checks for this, mcrypt is not an official Typo3 required extension AFAIK. 

It is a shame that Typo3 does not have the ability to encrypt db data in the core. It seems like it would be an important addition to also have a official encrypt/decrypt utility.  I find even encrypting email-addresses in DB would be a prudent habit. For example for BE_Users. If the DB gets compromised so do the email addresses.

I know that it is also possible to encrypt/decrypt in MySQL.
https://dev.mysql.com/doc/refman/5.5/en/encryption-functions.html
 
However it is only 128-bit AES, to get 256-bit you need to recompile MySQL. 
I feel these days 256-bit is a minimum.

Something to think about.

File: Classes/Mailer.php [hunk 1]
```diff
@@ -144,7 +144,9 @@ private function setTitle($htmlSrc)
     {
         // Extract title from HTML
         preg_match('|<title[^>]*>(.*)</title>|i', $htmlSrc, $m);
-        $title = trim($m[1]);
+
+        // As this is being extracted from HTML and it is being used as an email subject we need to decode any entities.
+        $title = trim(html_entity_decode($m[1]));
```
Comment (PowerKiKi) [inline Classes/Mailer.php]: This should probably be another PR, but most importantly it should be covered with tests.

File: Classes/Tca/BounceAccountTca.php [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+ * @license http://www.gnu.org/licenses/gpl.html GNU General Public License, version 3 or later
+ */
+class BounceAccountTca
+{
+
+    /**
+     * Adds new JavaScript function for evaluation of the TCA fields in backend
+     * @return 	string		JavaScript
+     */
+    public function returnFieldJS()
```
Comment (PowerKiKi) [inline Classes/Tca/BounceAccountTca.php]: That seems useless and should be deleted

File: Classes/Tools.php [hunk 3]
```diff
@@ -33,7 +33,7 @@
  *
  * @license http://www.gnu.org/licenses/gpl.html GNU General Public License, version 3 or later
  */
-abstract class Tools
+class Tools
```
Comment (PowerKiKi) [inline Classes/Tools.php]: I'd rather keep this class abstract and move the update code in a dedicated update class. This could be a nice opportunity to get rid of old style `class.ext_update.php` and use the SignalSlot approach instead to instantiate a dedicated update class. I noticed users regularly forget to click on the update button anyway...

**Comment by PowerKiKi** [comment]
Existing tests are broken, please fix them, probably something in `Tests/Functional/Fixtures/fixtures.xml`. Also new code is not covered by tests and should be. That should include at least `BounceAccount` (getPort/setPort), `Tools` (crypt/decrypt), `Mailer` (email subject).

Documentation also need to be adapted to mention the new advanced mode and markers.

Finally declare the new dependency in `composer.json` (don't forget to run `composer update --lock` after that). Something like:
```
    "require": {
        "ext-mcrypt": "*",
    }
```

File: Classes/Mailer.php (hunk 1 above)
Comment (marvin-martian) [inline Classes/Mailer.php]: What is there to test???  The code is extracting the subject from HTML source (not my code) your test is making bad assumptions as you are not testing for this (this is a classic example of why I think coders should not write their own unit tests). If you are setting HTML content as a subject in an email it needs to be devoid of html_entities as it is not supported.

Covered in RFC's for mail headers:
https://tools.ietf.org/html/rfc2047
https://tools.ietf.org/html/rfc5890
https://tools.ietf.org/html/rfc6532

You are allowed to have entities in message bodies which are HTML.

File: Classes/Tools.php (hunk 3 above)
Comment (marvin-martian) [inline Classes/Tools.php]: You can do this, but it is inefficient, it just means the PHP has to parse another file. It is more efficient to not set it as abstract.  You use the Tools class throughout your script even though not once have I seen this abstract extended to another class as all the methods are static. So there is no reason for the class to be an abstract.

**Comment by marvin-martian** [comment]
> Also new code is not covered by tests and should be. 

Not it shouldn't. @see #75. But you are welcome to write them if it makes you happy but it is bit pointless writing tests for working code after the fact. I have this in production already on a very busy server, it works.

> Documentation also need to be adapted to mention the new advanced mode and markers.

Agreed, I will look into adding some.

> Finally declare the new dependency in composer.json

No problems.

File: Classes/Tca/BounceAccountTca.php (hunk 2 above)
Comment (marvin-martian) [inline Classes/Tca/BounceAccountTca.php]: You are probably right.

File: Classes/Tools.php (hunk 3 above)
Comment (marvin-martian) [inline Classes/Tools.php]: Ahh sorry I see what you mean... yes. lets do that in an Update class. much cleaner. I will make changes.

**Comment by marvin-martian** [comment]
BTW I have a bunch of other fixes but I will stagger them in separate PR. I have RealURL working cleanly which required some more moderate changes across the entire app.

**Comment by marvin-martian** [comment]
I had out of the bag for another extension I am working on a method for handling extension updates. Basically it makes the extension writer just focus on the task of the update and not how it displays in the BE. It also makes the coder think about implementing non-destructive updates. So when for some reason your update fails it tries to roll-back the changes if possible to not destroy the users data. It is not entirely complete I have not finished transacted file-operations yet. I haven't need to do one so it will come on a as needed basis.

Take a look, I think it is a nice idea, so many times I have had extensions make a mess of updates.  It also keeps track of the update history of the extension and lets you set your task as automatic, manual or both. Eventually I would like to get something like this as standard way via extbase.

**Comment by PowerKiKi** [comment]
It seems you created a PR from your branch `master` instead of a dedicated branch, which has the unfortunate effect that all commits to your fork appears in this PR. That makes it very difficult to review and merge (27 commits and 134 files changed on at least 5 different topics!). Would you be willing to create separate PR (branches) for each distinct topics please ? as it is now I won't be able to merge anything.

**Comment by marvin-martian** [comment]
Fair enough, sorry I was in the flow when I start pumping code. I will split this up for you and close this one down. I thought about it when I did the first push to master, but I was on a roll and thought... ahh f*ck it ...and kept going. lol. sorry I was having a tough week and I was venting by coding, which is a nice outlet. :-)

**Comment by PowerKiKi** [comment]
no worries, take your time, there is no rush here ;-)
